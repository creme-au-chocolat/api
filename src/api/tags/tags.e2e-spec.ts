jest.mock('./tags.service');

import { faker } from '@faker-js/faker';
import { INestApplication } from '@nestjs/common';
import { getModelToken } from '@nestjs/mongoose';
import { Test } from '@nestjs/testing';
import * as request from 'supertest';
import { createTestingApp, testRequests } from '../../../test/utils/e2e';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import { TagDocument } from '../../shared/schemas/tag.schema';
import { TagsModule } from './tags.module';
import { TagsService } from './tags.service';
import {
  mockedTags,
  TagsService as MockedTagsService,
} from './__mocks__/tags.service';

const { helpers } = faker;

describe('TagsController (e2e)', () => {
  let app: INestApplication;
  let tagsService: TagsService;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [TagsModule],
    })
      .overrideProvider(TagsService)
      .useClass(MockedTagsService)
      .overrideProvider(getModelToken(TagDocument.name))
      .useValue({})
      .compile();

    tagsService = moduleRef.get(TagsService);

    app = createTestingApp(moduleRef);
    await app.init();
  });

  beforeEach(async () => {
    faker.seed(0);
  });

  describe('/tags/list/:category (GET)', () => {
    it('sort by popularity', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/list/artists?popular=true&page=1')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('sort by name', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/list/artists?popular=false&page=1')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('throws 404 error when requested page does not exist', async () => {
      const numberOfPages = await tagsService.getPageCount(CATEGORIES.artists);

      return request(app.getHttpServer())
        .get(`/tags/list/artists?popular=false&page=${numberOfPages + 1}`)
        .expect(404);
    });

    it('throws 400 error on wrong parameters values', async () => {
      const httpServer = app.getHttpServer();
      const badRequests: string[] = [
        '/tags/list/notacategory?popular=false&page=1',
        '/tags/list/artists?popular=false&page=0',
        '/tags/list/artists?popular=false&page=1.50',
        '/tags/list/artists?popular=false&page=notanumber',
        '/tags/list/artists?popular=false&page=-1',
      ];

      await testRequests(httpServer, badRequests, 400);
    });

    it('converts parameters types', async () => {
      const httpServer = app.getHttpServer();

      const shouldOrderByPopularity: string[] = [
        '/tags/list/artists?popular=true&page=1',
        '/tags/list/artists?popular=1&page=1',
        '/tags/list/artists?popular=anything&page=1',
      ];
      const shouldOrderByName: string[] = [
        '/tags/list/artists?popular=false&page=1',
        '/tags/list/artists?page=1',
      ];
      const shouldWorks: string[] = [
        '/tags/list/artists',
        '/tags/list/artists?page=1',
      ];

      for (const req of shouldOrderByPopularity) {
        const response = await request(httpServer)
          .get(req)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      }

      for (const req of shouldOrderByName) {
        const response = await request(httpServer)
          .get(req)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      }

      for (const req of shouldWorks) {
        await request(httpServer)
          .get(req)
          .expect(200);
      }
    });
  });

  describe('/tags/list/:category/:letter (GET)', () => {
    it('returns all tags starting with specified letter, sorted by name', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/list/artists/a')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('is case insensitive', async () => {
      const lowercaseResponse = await request(app.getHttpServer())
        .get('/tags/list/artists/a')
        .expect(200);

      const uppercaseResponse = await request(app.getHttpServer())
        .get('/tags/list/artists/A')
        .expect(200);

      expect(lowercaseResponse.body).toMatchSnapshot();
      expect(uppercaseResponse.body).toMatchSnapshot();
      expect(uppercaseResponse.body).toStrictEqual(lowercaseResponse.body);
    });

    it('throws 400 error on wrong parameters values', async () => {
      const httpServer = app.getHttpServer();
      const badRequests: string[] = [
        '/tags/list/notacategory/a',
        '/tags/list/artists/notaletter',
      ];

      await testRequests(httpServer, badRequests, 400);
    });
  });

  describe('/tags/:id/details (GET)', () => {
    it('returns requested tag', async () => {
      const randomTag = helpers.arrayElement(mockedTags);

      const response = await request(app.getHttpServer())
        .get(`/tags/${randomTag.id}/details`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('returns 404 error when tag does not exist', async () => {
      await request(app.getHttpServer())
        .get('/tags/999999/details')
        .expect(404);
    });

    it('validate and convert parameters', async () => {
      const httpServer = app.getHttpServer();
      const badRequests = [
        '/tags/0/details',
        '/tags/bite/details',
        '/tags/NaN/details',
      ];
      const goodRequests = [
        `/tags/${helpers.arrayElement(mockedTags).id}/details`,
      ];

      await testRequests(httpServer, badRequests, 400);
      await testRequests(httpServer, goodRequests, 200);
    });
  });

  describe('/tags/search (GET)', () => {
    it('returns tags including search query', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/search?q=ips')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('returns tags including search query and category', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/search?q=ips&category=tags')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('returns empty array if nothing found', async () => {
      await request(app.getHttpServer())
        .get('/tags/search?q=impossiblestring')
        .expect(200)
        .expect([]);
    });

    it('validate and convert parameters', async () => {
      const httpServer = app.getHttpServer();
      const badRequests = [
        '/tags/search',
        '/tags/search?q=aa',
        '/tags/search?q=aaa&page=0',
        '/tags/search?q=aaa&page=1.3',
      ];
      const goodRequests = [
        `/tags/search?q=aaa&page=1&category=artists`,
        `/tags/search?q=aaa`,
      ];

      await testRequests(httpServer, badRequests, 400);
      await testRequests(httpServer, goodRequests, 200);
    });
  });

  afterAll(async () => {
    await app.close();
  });
});
