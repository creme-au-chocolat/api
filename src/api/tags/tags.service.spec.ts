import { faker } from '@faker-js/faker';
import { CacheModule, NotFoundException } from '@nestjs/common';
import { getConnectionToken } from '@nestjs/mongoose';
import { Test } from '@nestjs/testing';
import { Connection } from 'mongoose';
import { Tag } from 'src/shared/types/tag.entity';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import {
  closeMongoConnection,
  DatabaseTestingModule,
} from '../../testing/database-testing/database-testing.module';
import { SeederService } from '../../testing/database-testing/seeder/seeder.service';
import { TagsService } from './tags.service';

const { helpers } = faker;

describe('CategoriesController', () => {
  let tagsService: TagsService;
  let databaseConnection: Connection;
  let seededWith: Tag[];

  beforeEach(async () => {
    faker.seed(1);

    TagsService.PAGE_SIZE = 10;

    const moduleRef = await Test.createTestingModule({
      controllers: [],
      providers: [TagsService],
      imports: [CacheModule.register({ ttl: 1 }), DatabaseTestingModule],
    }).compile();

    tagsService = moduleRef.get(TagsService);
    databaseConnection = moduleRef.get(getConnectionToken());

    const seedingService = moduleRef.get(SeederService);
    [, seededWith] = await seedingService.seedTags();
  });

  afterEach(async () => {
    await databaseConnection.close();
    await closeMongoConnection();
  });

  describe('getPageCount', () => {
    it('returns correct page count', async () => {
      expect(
        await tagsService.getPageCount(CATEGORIES.artists),
      ).toMatchSnapshot();
    });
  });

  describe('getTagsByPopularity', () => {
    it('returns tags sorted by popularity', async () => {
      await expect(
        tagsService.getTagsByPopularity(CATEGORIES.artists, 1),
      ).resolves.toMatchSnapshot();
    });

    it('works with other pages', async () => {
      await expect(
        tagsService.getTagsByPopularity(CATEGORIES.artists, 2),
      ).resolves.toMatchSnapshot();
    });

    it('returns empty array when page out of range', async () => {
      await expect(
        tagsService.getTagsByPopularity(CATEGORIES.artists, 100000),
      ).resolves.toEqual([]);
    });
  });

  describe('getTagsByLetter', () => {
    it('returns all tags starting with requested letter', async () => {
      await expect(
        tagsService.getTagsByLetter(CATEGORIES.artists, 'A'),
      ).resolves.toMatchSnapshot();
    });

    it('is case insensitive', async () => {
      await expect(
        tagsService.getTagsByLetter(CATEGORIES.artists, 'A'),
      ).resolves.toMatchSnapshot();

      await expect(
        tagsService.getTagsByLetter(CATEGORIES.artists, 'a'),
      ).resolves.toMatchSnapshot();
    });

    it('returns empty array when letter is invalid', async () => {
      await expect(
        tagsService.getTagsByLetter(CATEGORIES.artists, '$'),
      ).resolves.toEqual([]);
    });
  });

  describe('getTags', () => {
    it('returns all tags sorted by name', async () => {
      await expect(
        tagsService.getTags(CATEGORIES.artists, 1),
      ).resolves.toMatchSnapshot();
    });

    it('works with any page', async () => {
      await expect(
        tagsService.getTags(CATEGORIES.artists, 2),
      ).resolves.toMatchSnapshot();
    });

    it('returns empty array when page out of range', async () => {
      await expect(
        tagsService.getTags(CATEGORIES.artists, 100000),
      ).resolves.toEqual([]);
    });
  });

  describe('search', () => {
    it('returns all tags matching search query', async () => {
      await expect(tagsService.search('i', 1)).resolves.toMatchSnapshot();
    });

    it('returns all tags matching search query and category', async () => {
      await expect(
        tagsService.search('i', 1, CATEGORIES.artists),
      ).resolves.toMatchSnapshot();
    });

    it('works with any page', async () => {
      await expect(tagsService.search('i', 2)).resolves.toMatchSnapshot();
    });
  });

  describe('getTagById', () => {
    it('returns requested tag', async () => {
      const randomTag = helpers.arrayElement(seededWith);

      await expect(
        tagsService.getTagById(randomTag.id),
      ).resolves.toMatchSnapshot();
    });

    it('throw NotFoundException when tag does not exist', async () => {
      await expect(
        tagsService.getTagById(Number.POSITIVE_INFINITY),
      ).rejects.toThrowError(NotFoundException);
    });
  });
});
