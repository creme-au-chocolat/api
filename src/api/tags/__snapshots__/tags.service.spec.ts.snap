// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesController getPageCount returns correct page count 1`] = `3`;

exports[`CategoriesController getTagById returns requested tag 1`] = `
Object {
  "category": "artists",
  "id": 203906,
  "name": "dignissimos",
  "tagged": 55552,
  "uri": "https://far-working.info",
}
`;

exports[`CategoriesController getTags returns all tags sorted by name 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 828844,
    "name": "autem",
    "tagged": 25470,
    "uri": "https://boiling-obesity.com",
  },
  Object {
    "category": "artists",
    "id": 39054,
    "name": "debitis",
    "tagged": 86502,
    "uri": "https://svelte-bit.net",
  },
  Object {
    "category": "artists",
    "id": 203906,
    "name": "dignissimos",
    "tagged": 55552,
    "uri": "https://far-working.info",
  },
  Object {
    "category": "artists",
    "id": 456974,
    "name": "dolore",
    "tagged": 22539,
    "uri": "http://prestigious-rhinoceros.org",
  },
  Object {
    "category": "artists",
    "id": 789278,
    "name": "doloribus",
    "tagged": 66063,
    "uri": "https://focused-downgrade.info",
  },
  Object {
    "category": "artists",
    "id": 842030,
    "name": "eos",
    "tagged": 28871,
    "uri": "https://elliptical-meeting.org",
  },
  Object {
    "category": "artists",
    "id": 66536,
    "name": "et",
    "tagged": 7425,
    "uri": "http://bruised-rip.com",
  },
  Object {
    "category": "artists",
    "id": 450873,
    "name": "et",
    "tagged": 466,
    "uri": "http://fair-plier.info",
  },
  Object {
    "category": "artists",
    "id": 731727,
    "name": "et",
    "tagged": 70433,
    "uri": "http://odd-satisfaction.org",
  },
  Object {
    "category": "artists",
    "id": 854114,
    "name": "et",
    "tagged": 62463,
    "uri": "https://disgusting-song.name",
  },
  Object {
    "category": "artists",
    "id": 798603,
    "name": "inventore",
    "tagged": 87985,
    "uri": "https://keen-councilman.net",
  },
  Object {
    "category": "artists",
    "id": 370083,
    "name": "libero",
    "tagged": 46215,
    "uri": "http://competent-baritone.net",
  },
  Object {
    "category": "artists",
    "id": 22330,
    "name": "modi",
    "tagged": 4830,
    "uri": "https://unrealistic-self-control.net",
  },
  Object {
    "category": "artists",
    "id": 967275,
    "name": "natus",
    "tagged": 65432,
    "uri": "http://colossal-battleship.biz",
  },
  Object {
    "category": "artists",
    "id": 878831,
    "name": "nisi",
    "tagged": 28147,
    "uri": "https://grimy-mozzarella.org",
  },
  Object {
    "category": "artists",
    "id": 546346,
    "name": "nulla",
    "tagged": 2388,
    "uri": "https://white-influence.info",
  },
  Object {
    "category": "artists",
    "id": 949015,
    "name": "perferendis",
    "tagged": 42681,
    "uri": "http://second-orient.net",
  },
  Object {
    "category": "artists",
    "id": 499583,
    "name": "quaerat",
    "tagged": 21868,
    "uri": "http://insignificant-concrete.info",
  },
  Object {
    "category": "artists",
    "id": 599109,
    "name": "quibusdam",
    "tagged": 57706,
    "uri": "http://coarse-kinase.org",
  },
  Object {
    "category": "artists",
    "id": 699757,
    "name": "sunt",
    "tagged": 61393,
    "uri": "http://well-groomed-needle.biz",
  },
  Object {
    "category": "artists",
    "id": 181612,
    "name": "vel",
    "tagged": 70568,
    "uri": "http://energetic-win.com",
  },
  Object {
    "category": "artists",
    "id": 540951,
    "name": "vel",
    "tagged": 50299,
    "uri": "https://speedy-library.org",
  },
]
`;

exports[`CategoriesController getTags works with any page 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 798603,
    "name": "inventore",
    "tagged": 87985,
    "uri": "https://keen-councilman.net",
  },
  Object {
    "category": "artists",
    "id": 370083,
    "name": "libero",
    "tagged": 46215,
    "uri": "http://competent-baritone.net",
  },
  Object {
    "category": "artists",
    "id": 22330,
    "name": "modi",
    "tagged": 4830,
    "uri": "https://unrealistic-self-control.net",
  },
  Object {
    "category": "artists",
    "id": 967275,
    "name": "natus",
    "tagged": 65432,
    "uri": "http://colossal-battleship.biz",
  },
  Object {
    "category": "artists",
    "id": 878831,
    "name": "nisi",
    "tagged": 28147,
    "uri": "https://grimy-mozzarella.org",
  },
  Object {
    "category": "artists",
    "id": 546346,
    "name": "nulla",
    "tagged": 2388,
    "uri": "https://white-influence.info",
  },
  Object {
    "category": "artists",
    "id": 949015,
    "name": "perferendis",
    "tagged": 42681,
    "uri": "http://second-orient.net",
  },
  Object {
    "category": "artists",
    "id": 499583,
    "name": "quaerat",
    "tagged": 21868,
    "uri": "http://insignificant-concrete.info",
  },
  Object {
    "category": "artists",
    "id": 599109,
    "name": "quibusdam",
    "tagged": 57706,
    "uri": "http://coarse-kinase.org",
  },
  Object {
    "category": "artists",
    "id": 699757,
    "name": "sunt",
    "tagged": 61393,
    "uri": "http://well-groomed-needle.biz",
  },
  Object {
    "category": "artists",
    "id": 181612,
    "name": "vel",
    "tagged": 70568,
    "uri": "http://energetic-win.com",
  },
  Object {
    "category": "artists",
    "id": 540951,
    "name": "vel",
    "tagged": 50299,
    "uri": "https://speedy-library.org",
  },
]
`;

exports[`CategoriesController getTagsByLetter is case insensitive 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 828844,
    "name": "autem",
    "tagged": 25470,
    "uri": "https://boiling-obesity.com",
  },
]
`;

exports[`CategoriesController getTagsByLetter is case insensitive 2`] = `
Array [
  Object {
    "category": "artists",
    "id": 828844,
    "name": "autem",
    "tagged": 25470,
    "uri": "https://boiling-obesity.com",
  },
]
`;

exports[`CategoriesController getTagsByLetter returns all tags starting with requested letter 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 828844,
    "name": "autem",
    "tagged": 25470,
    "uri": "https://boiling-obesity.com",
  },
]
`;

exports[`CategoriesController getTagsByPopularity returns tags sorted by popularity 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 798603,
    "name": "inventore",
    "tagged": 87985,
    "uri": "https://keen-councilman.net",
  },
  Object {
    "category": "artists",
    "id": 39054,
    "name": "debitis",
    "tagged": 86502,
    "uri": "https://svelte-bit.net",
  },
  Object {
    "category": "artists",
    "id": 181612,
    "name": "vel",
    "tagged": 70568,
    "uri": "http://energetic-win.com",
  },
  Object {
    "category": "artists",
    "id": 731727,
    "name": "et",
    "tagged": 70433,
    "uri": "http://odd-satisfaction.org",
  },
  Object {
    "category": "artists",
    "id": 789278,
    "name": "doloribus",
    "tagged": 66063,
    "uri": "https://focused-downgrade.info",
  },
  Object {
    "category": "artists",
    "id": 967275,
    "name": "natus",
    "tagged": 65432,
    "uri": "http://colossal-battleship.biz",
  },
  Object {
    "category": "artists",
    "id": 854114,
    "name": "et",
    "tagged": 62463,
    "uri": "https://disgusting-song.name",
  },
  Object {
    "category": "artists",
    "id": 699757,
    "name": "sunt",
    "tagged": 61393,
    "uri": "http://well-groomed-needle.biz",
  },
  Object {
    "category": "artists",
    "id": 599109,
    "name": "quibusdam",
    "tagged": 57706,
    "uri": "http://coarse-kinase.org",
  },
  Object {
    "category": "artists",
    "id": 203906,
    "name": "dignissimos",
    "tagged": 55552,
    "uri": "https://far-working.info",
  },
  Object {
    "category": "artists",
    "id": 540951,
    "name": "vel",
    "tagged": 50299,
    "uri": "https://speedy-library.org",
  },
  Object {
    "category": "artists",
    "id": 370083,
    "name": "libero",
    "tagged": 46215,
    "uri": "http://competent-baritone.net",
  },
  Object {
    "category": "artists",
    "id": 949015,
    "name": "perferendis",
    "tagged": 42681,
    "uri": "http://second-orient.net",
  },
  Object {
    "category": "artists",
    "id": 842030,
    "name": "eos",
    "tagged": 28871,
    "uri": "https://elliptical-meeting.org",
  },
  Object {
    "category": "artists",
    "id": 878831,
    "name": "nisi",
    "tagged": 28147,
    "uri": "https://grimy-mozzarella.org",
  },
  Object {
    "category": "artists",
    "id": 828844,
    "name": "autem",
    "tagged": 25470,
    "uri": "https://boiling-obesity.com",
  },
  Object {
    "category": "artists",
    "id": 456974,
    "name": "dolore",
    "tagged": 22539,
    "uri": "http://prestigious-rhinoceros.org",
  },
  Object {
    "category": "artists",
    "id": 499583,
    "name": "quaerat",
    "tagged": 21868,
    "uri": "http://insignificant-concrete.info",
  },
  Object {
    "category": "artists",
    "id": 66536,
    "name": "et",
    "tagged": 7425,
    "uri": "http://bruised-rip.com",
  },
  Object {
    "category": "artists",
    "id": 22330,
    "name": "modi",
    "tagged": 4830,
    "uri": "https://unrealistic-self-control.net",
  },
  Object {
    "category": "artists",
    "id": 546346,
    "name": "nulla",
    "tagged": 2388,
    "uri": "https://white-influence.info",
  },
  Object {
    "category": "artists",
    "id": 450873,
    "name": "et",
    "tagged": 466,
    "uri": "http://fair-plier.info",
  },
]
`;

exports[`CategoriesController getTagsByPopularity works with other pages 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 540951,
    "name": "vel",
    "tagged": 50299,
    "uri": "https://speedy-library.org",
  },
  Object {
    "category": "artists",
    "id": 370083,
    "name": "libero",
    "tagged": 46215,
    "uri": "http://competent-baritone.net",
  },
  Object {
    "category": "artists",
    "id": 949015,
    "name": "perferendis",
    "tagged": 42681,
    "uri": "http://second-orient.net",
  },
  Object {
    "category": "artists",
    "id": 842030,
    "name": "eos",
    "tagged": 28871,
    "uri": "https://elliptical-meeting.org",
  },
  Object {
    "category": "artists",
    "id": 878831,
    "name": "nisi",
    "tagged": 28147,
    "uri": "https://grimy-mozzarella.org",
  },
  Object {
    "category": "artists",
    "id": 828844,
    "name": "autem",
    "tagged": 25470,
    "uri": "https://boiling-obesity.com",
  },
  Object {
    "category": "artists",
    "id": 456974,
    "name": "dolore",
    "tagged": 22539,
    "uri": "http://prestigious-rhinoceros.org",
  },
  Object {
    "category": "artists",
    "id": 499583,
    "name": "quaerat",
    "tagged": 21868,
    "uri": "http://insignificant-concrete.info",
  },
  Object {
    "category": "artists",
    "id": 66536,
    "name": "et",
    "tagged": 7425,
    "uri": "http://bruised-rip.com",
  },
  Object {
    "category": "artists",
    "id": 22330,
    "name": "modi",
    "tagged": 4830,
    "uri": "https://unrealistic-self-control.net",
  },
  Object {
    "category": "artists",
    "id": 546346,
    "name": "nulla",
    "tagged": 2388,
    "uri": "https://white-influence.info",
  },
  Object {
    "category": "artists",
    "id": 450873,
    "name": "et",
    "tagged": 466,
    "uri": "http://fair-plier.info",
  },
]
`;

exports[`CategoriesController search returns all tags matching search query 1`] = `
Array [
  Object {
    "category": "tags",
    "id": 988615,
    "name": "id",
    "tagged": 76257,
    "uri": "https://jittery-estuary.name",
  },
  Object {
    "category": "parodies",
    "id": 931971,
    "name": "id",
    "tagged": 52273,
    "uri": "https://serpentine-timber.com",
  },
  Object {
    "category": "parodies",
    "id": 586832,
    "name": "in",
    "tagged": 41738,
    "uri": "https://granular-drizzle.net",
  },
  Object {
    "category": "artists",
    "id": 798603,
    "name": "inventore",
    "tagged": 87985,
    "uri": "https://keen-councilman.net",
  },
  Object {
    "category": "tags",
    "id": 184010,
    "name": "ipsam",
    "tagged": 23526,
    "uri": "https://stupid-mortgage.name",
  },
  Object {
    "category": "parodies",
    "id": 421107,
    "name": "ipsum",
    "tagged": 5924,
    "uri": "https://querulous-bottom-line.name",
  },
  Object {
    "category": "groups",
    "id": 121343,
    "name": "iste",
    "tagged": 42838,
    "uri": "http://grubby-baggy.com",
  },
  Object {
    "category": "artists",
    "id": 203906,
    "name": "dignissimos",
    "tagged": 55552,
    "uri": "https://far-working.info",
  },
  Object {
    "category": "groups",
    "id": 641565,
    "name": "hic",
    "tagged": 92520,
    "uri": "https://dreary-topsail.com",
  },
  Object {
    "category": "artists",
    "id": 370083,
    "name": "libero",
    "tagged": 46215,
    "uri": "http://competent-baritone.net",
  },
]
`;

exports[`CategoriesController search returns all tags matching search query and category 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 798603,
    "name": "inventore",
    "tagged": 87985,
    "uri": "https://keen-councilman.net",
  },
  Object {
    "category": "artists",
    "id": 203906,
    "name": "dignissimos",
    "tagged": 55552,
    "uri": "https://far-working.info",
  },
  Object {
    "category": "artists",
    "id": 370083,
    "name": "libero",
    "tagged": 46215,
    "uri": "http://competent-baritone.net",
  },
  Object {
    "category": "artists",
    "id": 878831,
    "name": "nisi",
    "tagged": 28147,
    "uri": "https://grimy-mozzarella.org",
  },
  Object {
    "category": "artists",
    "id": 599109,
    "name": "quibusdam",
    "tagged": 57706,
    "uri": "http://coarse-kinase.org",
  },
  Object {
    "category": "artists",
    "id": 39054,
    "name": "debitis",
    "tagged": 86502,
    "uri": "https://svelte-bit.net",
  },
  Object {
    "category": "artists",
    "id": 22330,
    "name": "modi",
    "tagged": 4830,
    "uri": "https://unrealistic-self-control.net",
  },
  Object {
    "category": "artists",
    "id": 789278,
    "name": "doloribus",
    "tagged": 66063,
    "uri": "https://focused-downgrade.info",
  },
  Object {
    "category": "artists",
    "id": 949015,
    "name": "perferendis",
    "tagged": 42681,
    "uri": "http://second-orient.net",
  },
]
`;

exports[`CategoriesController search works with any page 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 878831,
    "name": "nisi",
    "tagged": 28147,
    "uri": "https://grimy-mozzarella.org",
  },
  Object {
    "category": "tags",
    "id": 743825,
    "name": "similique",
    "tagged": 27464,
    "uri": "http://black-and-white-descent.com",
  },
  Object {
    "category": "parodies",
    "id": 68209,
    "name": "similique",
    "tagged": 6703,
    "uri": "http://strong-fiberglass.name",
  },
  Object {
    "category": "parodies",
    "id": 621805,
    "name": "sint",
    "tagged": 79037,
    "uri": "http://dry-cheetah.biz",
  },
  Object {
    "category": "tags",
    "id": 246210,
    "name": "sit",
    "tagged": 39631,
    "uri": "http://carefree-android.name",
  },
  Object {
    "category": "parodies",
    "id": 846827,
    "name": "adipisci",
    "tagged": 93871,
    "uri": "https://happy-go-lucky-wash.com",
  },
  Object {
    "category": "parodies",
    "id": 137135,
    "name": "eligendi",
    "tagged": 55132,
    "uri": "https://big-hearted-cloud.info",
  },
  Object {
    "category": "tags",
    "id": 964839,
    "name": "enim",
    "tagged": 35653,
    "uri": "https://silky-highlight.org",
  },
  Object {
    "category": "characters",
    "id": 685218,
    "name": "qui",
    "tagged": 93553,
    "uri": "https://testy-head.biz",
  },
  Object {
    "category": "parodies",
    "id": 237026,
    "name": "qui",
    "tagged": 35531,
    "uri": "https://warm-hacienda.info",
  },
]
`;
